README.md
# imgurtv
Check out the anouncement on Imgur: http://imgur.com/a/MQvyE


plugin.picture.imgur/changelog.txt
v0.0.1
 - Initial Release (2015-06-15)


plugin.picture.imgur/addon.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
# Aw motherfuckin breadcrumbs, oh YISS!
#  _
# (_)_ __ ___   __ _ _   _ _ __ 
# | | '_ ` _ \ / _` | | | | '__|
# | | | | | | | (_| | |_| | |   
# |_|_| |_| |_|\__, |\__,_|_|   
#              |___/  
# 
# Author: Andre Oosthuizen
# Date: 2015-06-14
#
import sys
import urllib
import urlparse
import xbmc
import xbmcaddon
import xbmcgui
import xbmcplugin
import resources.lib.imgurpython.client as client
import resources.lib.imgurpython.helpers.error as error

baseUrl = sys.argv[0]
addonHandle = int(sys.argv[1])
args = urlparse.parse_qs(sys.argv[2][1:])

#TODO Settings
addon = xbmcaddon.Addon()
pageSizeSetting = addon.getSetting("per_page")

client_id = '716f6289bf2413c'
client_secret = 'fdb0ba64d591f639907bc78ab2a926fedf05ce84'
client = client.ImgurClient(client_id, client_secret)

def createUrl(query):
    return baseUrl + '?' + urllib.urlencode(query)

def renderMenu():
    url = createUrl({'mode': 'renderGallery', 'section': 'hot'})
    menuItem1 = xbmcgui.ListItem('Most Viral', iconImage='DefaultFolder.png')
    xbmcplugin.addDirectoryItem(handle=addonHandle, url=url, listitem=menuItem1, isFolder=True)

    url = createUrl({'mode': 'renderGallery', 'section': 'user'})
    menuItem2 = xbmcgui.ListItem('User Submitted', iconImage='DefaultFolder.png')
    xbmcplugin.addDirectoryItem(handle=addonHandle, url=url, listitem=menuItem2, isFolder=True)

    xbmcplugin.endOfDirectory(addonHandle)
    
def renderGallery(section, page):
    try:
        galleryItems = client.gallery(section=section, sort='time', page=page, window='day', show_viral=True)
        for galleryItem in galleryItems:
            url = createUrl({'mode': 'renderGalleryItem', 'section': section, 'itemId': galleryItem.id})
            thumbnail = None
            icon = None
            if galleryItem.is_album:
                thumbnail = 'http://i.imgur.com/' + galleryItem.cover + 'b.jpg'
                icon = 'http://i.imgur.com/' + galleryItem.cover + 's.jpg'
            else:
                thumbnail = 'http://i.imgur.com/' + galleryItem.id + 'b.jpg'
                icon = 'http://i.imgur.com/' + galleryItem.id + 's.jpg'
            
            li = xbmcgui.ListItem(label=galleryItem.title, label2=galleryItem.description, iconImage=icon, thumbnailImage=thumbnail)
            li.setMimeType('image/jpeg')
            xbmcplugin.addDirectoryItem(handle=addonHandle, url=url, listitem=li)
            
    except error.ImgurClientError as e:
        xbmcgui.Dialog().notification("Unable to fetch gallery", 'Code ' + e.status_code + ": " + e.error_message, xbmcgui.NOTIFICATION_ERROR)  # @UndefinedVariable
        xbmc.log('Error ' + e.status_code + ': ' + e.error_message, xbmc.LOGERROR)
    except error.ImgurClientRateLimitError as e:
        xbmcgui.Dialog().notification("Rate limit error", 'Code ' + e.status_code + ": " + e.error_message, xbmcgui.NOTIFICATION_ERROR)  # @UndefinedVariable
        xbmc.log('Error ' + e.status_code + ': ' + e.error_message, xbmc.LOGERROR)
 
    xbmcplugin.endOfDirectory(addonHandle)
    #Change to thumbnail mode
    xbmc.executebuiltin('Container.SetViewMode(500)')


#
#  Banana for scale
#  _
# //\
# V  \
# \  \_
#   \,'.`-.
#    |\ `. `.       
#    ( \  `. `-.                        _,.-:\
#     \ \   `.  `-._             __..--' ,-';/
#      \ `.   `-.   `-..___..---'   _.--' ,'/
#       `. `.    `-._        __..--'    ,' /
#         `. `-_     ``--..''       _.-' ,'
#           `-_ `-.___        __,--'   ,'
#              `-.__  `----"""    __.-'
#                   `--..____..--'
#
mode = args.get('mode', None)
if mode is None:
    renderMenu()
elif mode[0] == 'renderGallery':
    section = args['section'][0]
    #TODO Add next page ListItem
    page = 0
    renderGallery(section=section, page=page)
elif mode[0] == 'renderGalleryItem':
    section = args['section'][0]
    itemId = args['itemId'][0]
    query = {'itemId': itemId, 'section': section}
    scriptUrl = 'special://home/addons/script.module.imgur/default.py'
    xbmc.executebuiltin("XBMC.RunScript("+scriptUrl+","+section+","+itemId+")")


script.module.imgur/lib/imgurpython/navigator.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
# Aw motherfuckin breadcrumbs, oh YISS!
#  _
# (_)_ __ ___   __ _ _   _ _ __ 
# | | '_ ` _ \ / _` | | | | '__|
# | | | | | | | (_| | |_| | |   
# |_|_| |_| |_|\__, |\__,_|_|   
#              |___/  
# 
# Author: Andre Oosthuizen
# Date: 2015-06-14
#
import xbmc
import xbmcgui
import xbmcaddon
from collections import deque
from .client import ImgurClient
from .helpers import error

CLIENT_ID = '716f6289bf2413c'
CLIENT_SECRET = 'fdb0ba64d591f639907bc78ab2a926fedf05ce84'

class GalleryNavigator():
    
    _client = None
    _section = None
    _itemId = None
    _galleryItem = None
    _galleryItems = None
    _cachedGalleryItems = deque()
    _cacheSize = 10
    
    def __init__(self):
        self._client = ImgurClient(CLIENT_ID, CLIENT_SECRET)
        self._cacheSize = int(xbmcaddon.Addon(id='script.module.imgur').getSetting('historyCache'))
        
    def init(self, section, itemId=None):
        xbmc.log('GalleryNavigator.init(): itemId='+str(itemId)+' section='+str(section))
        self._itemId = itemId
        self._section = section
        if (itemId is None):
            self._itemId = self.getNextItemId()
        self.loadItem()
        
    def loadItem(self):
        xbmc.log('GalleryNavigator.loadItem(): '+str(self._itemId))
        if (self._itemId is None):
            self._galleryItem = None
            return
        
        cachedGalleryItem = self.getCachedItem(self._itemId)
        if (cachedGalleryItem is not None):
            self._galleryItem = cachedGalleryItem
            return self._galleryItem
        
        try:
            self._galleryItem = self._client.gallery_item(self._itemId)
            self.addCachedItem(self._galleryItem)
            return self._galleryItem
        except error.ImgurClientError as e:
            xbmcgui.Dialog().notification("Unable to fetch gallery item", 'Code ' + str(e.status_code) + ": " + e.error_message, xbmcgui.NOTIFICATION_ERROR)  # @UndefinedVariable
            xbmc.log('Error ' + str(e.status_code) + ': ' + e.error_message, xbmc.LOGERROR)
        except error.ImgurClientRateLimitError as e:
            xbmcgui.Dialog().notification("Rate limit error", 'Code ' + str(e.status_code) + ": " + e.error_message, xbmcgui.NOTIFICATION_ERROR)  # @UndefinedVariable
            xbmc.log('Error ' + str(e.status_code) + ': ' + e.error_message, xbmc.LOGERROR)
            
    def loadGallery(self):
        xbmc.log('GalleryNavigator.loadGallery()')
        try:
            self._galleryItems = self._client.gallery(section=self._section, sort='time', page=0, window='day', show_viral=True)
        except error.ImgurClientError as e:
            xbmcgui.Dialog().notification("Unable to fetch gallery item", 'Code ' + str(e.status_code) + ": " + e.error_message, xbmcgui.NOTIFICATION_ERROR)  # @UndefinedVariable
            xbmc.log('Error ' + str(e.status_code) + ': ' + e.error_message, xbmc.LOGERROR)
        except error.ImgurClientRateLimitError as e:
            xbmcgui.Dialog().notification("Rate limit error", 'Code ' + str(e.status_code) + ": " + e.error_message, xbmcgui.NOTIFICATION_ERROR)  # @UndefinedVariable
            xbmc.log('Error ' + str(e.status_code) + ': ' + e.error_message, xbmc.LOGERROR)
    
    def item(self):
        return self._galleryItem
    
    def next(self):
        xbmc.log('GalleryNavigator.next()')
        self._itemId = self.getNextItemId()
        self.loadItem()
        return self._galleryItem
    
    def previous(self):
        self._itemId = self.getPreviousItemId()
        self.loadItem()
        return self._galleryItem
            
    def getNextItemId(self):
        xbmc.log('GalleryNavigator.getNextItemId()')
        if (self._galleryItems is None):
            self.loadGallery();
        if (self._itemId is None):
            return self._galleryItems[0].id
        found = False
        for item in self._galleryItems:
            if (found == True):
                return item.id
            if (item.id == self._itemId):
                found = True
        return None
            
    def getPreviousItemId(self):
        xbmc.log('GalleryNavigator.getPreviousItemId()')
        if (self._galleryItems is None):
            self.loadGallery();
        previousId = None
        for item in self._galleryItems:
            if (item.id == self._itemId):
                return previousId
            previousId = item.id
        return None
    
    def getCachedItem(self, itemId):
        for galleryItem in self._cachedGalleryItems:
            if (galleryItem.id == itemId):
                xbmc.log('Found '+str(itemId)+' in cache')
                return galleryItem
        return None
    
    def addCachedItem(self, galleryItem):
        self._cachedGalleryItems.append(galleryItem)
        if len(self._cachedGalleryItems) > self._cacheSize:
            self._cachedGalleryItems.popleft()
            
    def getPlayIds(self):
        ids = list()
        if (self._galleryItem.is_album):
            for galleryItem in self._galleryItem.images:
                if (galleryItem['animated'] == True):
                    ids.append(galleryItem['id'])
        else:
            if (self._galleryItem.animated == True):
                ids.append(self._galleryItem.id)
        return ids


script.module.imgur/default.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
# Aw motherfuckin breadcrumbs, oh YISS!
#  _
# (_)_ __ ___   __ _ _   _ _ __ 
# | | '_ ` _ \ / _` | | | | '__|
# | | | | | | | (_| | |_| | |   
# |_|_| |_| |_|\__, |\__,_|_|   
#              |___/  
# 
# Author: Andre Oosthuizen
# Date: 2015-06-14
#
import sys
import math
import datetime
import ImageFont
import xbmc
import xbmcaddon
import xbmcgui
from lib.imgurpython import GalleryNavigator

ACTION_MOVE_LEFT = 1
ACTION_MOVE_RIGHT = 2
ACTION_MOVE_UP = 3
ACTION_MOVE_DOWN = 4
ACTION_PAGE_UP = 5
ACTION_PAGE_DOWN = 6
ACTION_PARENT_DIR = 9
ACTION_PREVIOUS_MENU = 10
ACTION_MOUSE_WHEEL_UP = 104
ACTION_MOUSE_WHEEL_DOWN = 105
ACTION_SELECT_ITEM = 7

LAYOUT_PADDING = 10
LAYOUT_LEFT_LIST_WIDTH = 630
LAYOUT_LEFT_FIRST_LABEL_ID = 104
LAYOUT_LEFT_LAST_LABEL_ID = 140
LAYOUT_LEFT_SCROLL_TOP_ID = 101
LAYOUT_LEFT_SCROLL_BOTTOM_ID = 144
LAYOUT_RIGHT_LIST_WIDTH = 610
LAYOUT_RIGHT_FIRST_LABEL_ID = 201
LAYOUT_RIGHT_LAST_LABEL_ID = 220

SCROLL_AMOUNT = 70

addon = xbmcaddon.Addon(id='script.module.imgur')
galleryNavigator = GalleryNavigator()

def init():
    xbmc.log('init()', xbmc.LOGDEBUG)
    if len(sys.argv) == 3:
        section = sys.argv[1]
        itemId = sys.argv[2]
        galleryNavigator.init(section, itemId)
    else:
        section = 'hot'
        galleryNavigator.init(section)
    
class ImgurViewer(xbmcgui.WindowXML):
    
    def onInit(self):
        xbmcgui.WindowXML.onInit(self)
        self.initLeft()
        self.initRight()
        
    def initLeft(self):
        galleryItem = galleryNavigator.item()
        y = LAYOUT_PADDING
        y = self.setLabelControl(LAYOUT_LEFT_SCROLL_TOP_ID, galleryItem.title, y, font='font14', multiline=False)
        
        line2 = datetime.datetime.fromtimestamp(galleryItem.datetime).strftime('%a, %b %d at %H:%M')
        if galleryItem.account_url is not None:
            line2 = "by " + galleryItem.account_url + " - " + line2
        y = self.setLabelControl(102, line2, y, font='font12', multiline=False)
        
        lblTitleId = LAYOUT_LEFT_FIRST_LABEL_ID
        imgControlId = lblTitleId + 1
        imgOverlayId = imgControlId + 1
        lblDescId = imgOverlayId + 1
        if (galleryItem.is_album):
            y = self.setLabelControl(103, galleryItem.description, y)
            rowItems = 4
            for albumImage in galleryItem.images:
                if lblTitleId > LAYOUT_LEFT_LAST_LABEL_ID:
                    break
                y = self.setLabelControl(lblTitleId, albumImage['title'], y)
                lblTitleId = lblTitleId + rowItems
                
                y = self.setImageControl(imgControlId, albumImage['id'], albumImage['link'], albumImage['width'], albumImage['height'], y)
                self.overlayPlayButton(imgControlId, imgOverlayId, albumImage['animated'])
                imgControlId = imgControlId + rowItems
                imgOverlayId = imgOverlayId + rowItems
                
                y = self.setLabelControl(lblDescId, albumImage['description'], y)
                lblDescId = lblDescId + rowItems
        else:
            y = self.setImageControl(imgControlId, galleryItem.id, galleryItem.link, galleryItem.width, galleryItem.height, y)
            y = self.setLabelControl(lblDescId, galleryItem.description, y)
            self.overlayPlayButton(imgControlId, imgOverlayId, galleryItem.animated)
                
        y = self.setLabelControl(LAYOUT_LEFT_SCROLL_BOTTOM_ID, self.formatNumber(galleryItem.score)+' points, '+self.formatNumber(galleryItem.views)+' views', y)
        
        
    def initRight(self):
        galleryItem = galleryNavigator.item()
        y = LAYOUT_PADDING
        y = self.setLabelControl(200, self.formatNumber(galleryItem.comment_count)+' comments sorted by best', y, padding=LAYOUT_PADDING * 2)
        
        lblAuthorId = LAYOUT_RIGHT_FIRST_LABEL_ID
        lblCommentId = lblAuthorId + 1
        authorPadding = LAYOUT_PADDING / 2
        commentPadding = LAYOUT_PADDING + authorPadding
        for comment in galleryItem.comment_preview:
            if lblCommentId > LAYOUT_RIGHT_LAST_LABEL_ID:
                break
            authorLine = comment['author'] + ' : ' + str(comment['points']) + ' points'
            y = self.setLabelControl(lblAuthorId, authorLine, y, font='font12', padding=authorPadding, multiline=False)
            lblAuthorId = lblAuthorId + 2
            y = self.setLabelControl(lblCommentId, comment['comment'], y, padding=commentPadding)
            lblCommentId = lblCommentId + 2
        
    def onAction(self, action):
        if action == ACTION_PREVIOUS_MENU:
            self.close()
        if action == ACTION_PARENT_DIR:
            self.removeControl(self.strAction)
        if (action == ACTION_PAGE_UP or action == ACTION_MOUSE_WHEEL_UP or action == ACTION_MOVE_UP):
            self.scroll(SCROLL_AMOUNT)
        if (action == ACTION_PAGE_DOWN or action == ACTION_MOUSE_WHEEL_DOWN or action == ACTION_MOVE_DOWN):
            self.scroll(-SCROLL_AMOUNT)
        if (action == ACTION_MOVE_RIGHT):
            self.next()
        if (action == ACTION_MOVE_LEFT):
            self.previous()
        if (action == ACTION_SELECT_ITEM):
            self.playAll();
        
    def scroll(self, amount):
        for controlId in range(LAYOUT_LEFT_SCROLL_TOP_ID, LAYOUT_LEFT_SCROLL_BOTTOM_ID + 1):
            x = self.getControl(controlId).getX();
            y = self.getControl(controlId).getY();
            if (controlId == LAYOUT_LEFT_SCROLL_TOP_ID and amount > 0 and y == LAYOUT_PADDING):
                return
            if (amount < 0):
                lastControl = self.getControl(LAYOUT_LEFT_SCROLL_BOTTOM_ID)
                if (lastControl.getY() + lastControl.getHeight() + LAYOUT_PADDING <= 720):
                    return
            self.getControl(controlId).setPosition(x, y + amount);
        
    def setLabelControl(self, controlId, label, y, font='font13', padding=LAYOUT_PADDING, multiline=True):
        if (label is not None):
            x = self.getControl(controlId).getX()
            width = self.getControl(controlId).getWidth()
            self.getControl(controlId).setLabel(label.encode('utf-8'))
            self.getControl(controlId).setPosition(x, y)
            self.getControl(controlId).setVisible(True)
            height = self.calculateLabelHeight(label, width, font, multiline)
            xbmc.log('setLabelControl controlId='+str(controlId)+', label='+label.encode('utf-8')+', x='+str(x)+', y='+str(y)+', height='+str(height)+', padding='+str(padding), xbmc.LOGDEBUG)
            y = y + height + padding
        else:
            xbmc.log('setLabelControl controlId='+str(controlId)+', visible=FALSE', xbmc.LOGDEBUG)
            self.getControl(controlId).setVisible(False)
        return y
    
    def calculateLabelHeight(self, text, width, font, multiline):
        fontSize = 20
        if (font == 'font12'):
            fontSize = 17
        if (font == 'font13'):
            fontSize = 20
        if (font == 'font14'):
            fontSize = 22
        fontPath = xbmc.translatePath(addon.getAddonInfo('path')+'/resources/skins/Default/fonts/Roboto-Regular.ttf')
        font = ImageFont.truetype (fontPath, fontSize)
        extent = font.getsize(text.encode('utf-8'))
        textWidth = extent[0]
        textHeight = extent[1] + 1
        newlines = text.count('\n')
        if ((textWidth < width and newlines == 0) or (multiline == False)):
            return textHeight
        estLines = int(math.ceil(textWidth / float(width)))
        height = (estLines + newlines) * (textHeight)
        return height
        
    def setImageControl(self, controlId, imageId, link, width, height, y, maxWidth=LAYOUT_LEFT_LIST_WIDTH):
        imgUrl = self.getImageUrl(imageId, link)
        self.getControl(controlId).setImage(imgUrl)
        x = LAYOUT_PADDING
        if (width > maxWidth):
            heightRatio = float(height) / float(width)
            height = int(heightRatio * float(maxWidth))
            width = maxWidth
        if (width < maxWidth):
            x = x + int(float(maxWidth - width) / float(2))
        self.getControl(controlId).setWidth(width)
        self.getControl(controlId).setHeight(height)
        self.getControl(controlId).setPosition(x, y)
        self.getControl(controlId).setVisible(True)
        xbmc.log('setImageControl controlId='+str(controlId)+', imgUrl='+str(imgUrl)+', x='+str(x)+', y='+str(y)+', width='+str(width)+', height='+str(height), xbmc.LOGDEBUG)
        y = y + height + LAYOUT_PADDING
        return y
    
    def redraw(self):
        for controlId in range(LAYOUT_LEFT_SCROLL_TOP_ID, LAYOUT_LEFT_SCROLL_BOTTOM_ID):
            self.getControl(controlId).setVisible(False)
        for controlId in range(LAYOUT_RIGHT_FIRST_LABEL_ID, LAYOUT_RIGHT_LAST_LABEL_ID):
            self.getControl(controlId).setVisible(False)
        self.onInit()
        
    def next(self):
        xbmc.executebuiltin('ActivateWindow(busydialog)')
        galleryItem = galleryNavigator.next()
        if (galleryItem is not None):
            self.redraw()
        else:
            self.close()
        xbmc.executebuiltin('Dialog.Close(busydialog)')
        
    def previous(self):
        xbmc.executebuiltin('ActivateWindow(busydialog)')
        galleryItem = galleryNavigator.previous()
        if (galleryItem is not None):
            self.redraw()
        else:
            self.close()
        xbmc.executebuiltin('Dialog.Close(busydialog)')
    
    def formatNumber(self, number):
        s = '%d' % number
        groups = []
        while s and s[-1].isdigit():
            groups.append(s[-3:])
            s = s[:-3]
        return s + ','.join(reversed(groups))
    
    def playAll(self):
        playIds = galleryNavigator.getPlayIds()
        if len(playIds) == 0:
            return
        playlist = xbmc.PlayList(1)
        playlist.clear()
        for playId in playIds: 
            title = galleryNavigator.item().title + ' - ' + str(1)
            listitem = xbmcgui.ListItem(title, thumbnailImage='http://i.imgur.com/' + playId + 'b.jpg') 
            playlist.add('http://i.imgur.com/' + playId + '.mp4', listitem)
        xbmc.Player().play(playlist)
    
    def getImageUrl(self, imageId, link):
        imageQualityOptions = {"320x320":"m", "640x640":"l", "1024x1024":"h"}
        imageQuality = imageQualityOptions[addon.getSetting('imageQuality')]
        imgUrl = 'http://i.imgur.com/'+imageId + imageQuality+'.'+link[-3:]
        return imgUrl
    
    def overlayPlayButton(self, imageControlId, overlayControlId, animated):
        overlayControl = self.getControl(overlayControlId)
        if (animated):
            imageControl = self.getControl(imageControlId)
            y = imageControl.getY() + (imageControl.getHeight() / 2) - (overlayControl.getHeight() / 2)
            x = (LAYOUT_LEFT_LIST_WIDTH / 2) - (overlayControl.getWidth() / 2)
            overlayControl.setPosition(x, y)
            overlayControl.setVisible(True)
        else:
            overlayControl.setVisible(False)

        
#  Banana for scale
#  _
# //\
# V  \
# \  \_
#   \,'.`-.
#    |\ `. `.       
#    ( \  `. `-.                        _,.-:\
#     \ \   `.  `-._             __..--' ,-';/
#      \ `.   `-.   `-..___..---'   _.--' ,'/
#       `. `.    `-._        __..--'    ,' /
#         `. `-_     ``--..''       _.-' ,'
#           `-_ `-.___        __,--'   ,'
#              `-.__  `----"""    __.-'
#                   `--..____..--'
#

xbmc.executebuiltin('ActivateWindow(busydialog)')
init()
addonPath = xbmc.translatePath(addon.getAddonInfo('path'))
imgurViewer = ImgurViewer('scriptImgurViewer.xml', addonPath)
xbmc.executebuiltin('Dialog.Close(busydialog)')
imgurViewer.doModal()
del imgurViewer
